truffle/src/main/java/org/jruby/truffle/nodes/rubinius/IOBufferPrimitiveNodes.java

```diff
@@ -0,0 +1,193 @@
+        }
+
+        @Specialization
+        public RubyBasicObject allocate(VirtualFrame frame, RubyClass classToAllocate) {
+            final RubyBasicObject ioBuffer = new RubyBasicObject(classToAllocate);
+            rubyWithSelf(frame, ioBuffer, "@write_synced = true");
+            rubyWithSelf(frame, ioBuffer, "@storage = storage", "storage",
+                    new RubiniusByteArray(getContext().getCoreLibrary().getByteArrayClass(), new ByteList(IOBUFFER_SIZE)));
+            rubyWithSelf(frame, ioBuffer, "@used = 0");
+            rubyWithSelf(frame, ioBuffer, "@start = 0");
+            rubyWithSelf(frame, ioBuffer, "@total = total", "total", IOBUFFER_SIZE);
+            return ioBuffer;
+        }
+
+    }
+
+    @RubiniusPrimitive(name = "iobuffer_unshift")
+    public static abstract class IOBufferUnshiftPrimitiveNode extends RubiniusPrimitiveNode {
+
+        public IOBufferUnshiftPrimitiveNode(RubyContext context, SourceSection sourceSection) {
+            super(context, sourceSection);
+        }
+
+        public IOBufferUnshiftPrimitiveNode(IOBufferUnshiftPrimitiveNode prev) {
+            super(prev);
+        }
+
+        @Specialization
+        public int unshift(VirtualFrame frame, RubyBasicObject ioBuffer, RubyString string, int startPosition) {
+            rubyWithSelf(frame, ioBuffer, "@write_synced = false");
+
+            int stringSize = string.getByteList().realSize() - startPosition;
+            final int usedSpace = (int) rubyWithSelf(frame, ioBuffer, "@used");
+            final int availableSpace = IOBUFFER_SIZE - usedSpace;
+
+            if(stringSize > availableSpace) {
+                stringSize = availableSpace;
+            }
+
+            ByteList storage = ((RubiniusByteArray) rubyWithSelf(frame, ioBuffer, "@storage")).getBytes();
+
+            // Data is copied here - can we do something COW?
+            System.arraycopy(string.getByteList().unsafeBytes(), startPosition, storage.getUnsafeBytes(), usedSpace, stringSize);
+
+            rubyWithSelf(frame, ioBuffer, "@used = used", "used", usedSpace + stringSize);
+
+            return stringSize;
+        }
+
+    }
+
+    @RubiniusPrimitive(name = "iobuffer_fill")
+    public static abstract class IOBufferFillPrimitiveNode extends RubiniusPrimitiveNode {
+
+        public IOBufferFillPrimitiveNode(RubyContext context, SourceSection sourceSection) {
+            super(context, sourceSection);
+        }
+
+        public IOBufferFillPrimitiveNode(IOBufferFillPrimitiveNode prev) {
+            super(prev);
+        }
+
+        @Specialization
+        public int fill(VirtualFrame frame, RubyBasicObject ioBuffer, RubyBasicObject io) {
+            int fd = (int) rubyWithSelf(frame, io, "@descriptor");
+
+            // TODO CS 21-Apr-15 allocating this buffer for each read is crazy
+            byte[] readBuffer = new byte[STACK_BUF_SZ];
+            int count = STACK_BUF_SZ;
+
+            if(left(frame, ioBuffer) < count) {
```

Comment:
Yes, I noticed it as well. I suppose it's because this is transliterated from C++ code, but we should have sth checking against this.